### DISCUSSION TITLE
How to use Isar in a FCM background isolate? Or, how to check if Isar is already open?

### DISCUSSION THREAD

**Discussion by tlp19** [isar/isar#256] [Help] [answered]
Hi, I am trying to make an app that receives messages from Firebase Cloud Messaging and stores the data payload in an Isar Database.

I first tried this with HiveDB but it didn't work, and as Isar has multi-isolate support, I thought it was the perfect package for the job.

This was easy to implement for foreground notification, but when trying to apply this when the app is in background or terminated is proving to be more difficult.

Here is my background message callback function:

```dart
///Callback for when FCM messages are received when the app is either in the background or terminated
Future<void> _firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  print("Handling a background message: ${message.messageId}");

  // Opening Isar
  final dir = await getApplicationSupportDirectory();
  final isar = await Isar.open(
    schemas: [MessageSchema],
    directory: dir.path,
  );

  // Storing the message in the DB 
  await DatabaseService()
      .addToMessagesDB(isar, Message.fromJson(message.data));
}
```

This works perfectly for the first message that arrives in the background: it gets stored correctly in the database. But when a second message arrives in the background, I get the following error:
```console
I/flutter ( 7588): FlutterFire Messaging: An error occurred in your background messaging handler:
I/flutter ( 7588): IsarError: Instance has already been opened.
```

I think this is because the isolate remembers that Isar has already been opened, and when the callback is executed for the second time, this leads to an error when trying to open Isar again.

I have therefore been trying to find a way to **only open Isar if it is not already open,** but I am struggling to do this.

My first attempt was to try and get the Isar instance:
```dart
  // Opening Isar
  var isar = Isar.getInstance("messages");

  print("before, isar is: $isar"); //PRINT 1

  if (isar == null) {
    print("Isar is not open, so opening it.");
    final dir = await getApplicationSupportDirectory();
    isar = await Isar.open(
      schemas: [MessageSchema],
      directory: dir.path,
    );
  }

  print("after, isar is: $isar"); //PRINT 2

  // Storing the message in the DB
  await DatabaseService().addToMessagesDB(isar, Message.fromJson(message.data));
```

But when receiving the second message, I get:
```console
I/flutter ( 7588): before, isar is: null
I/flutter ( 7588): Isar is not open, so opening it.
I/flutter ( 7588): FlutterFire Messaging: An error occurred in your background messaging handler:
I/flutter ( 7588): IsarError: Instance has already been opened.
```

So I've been trying to use `isar.isOpen` to check if Isar is actually open, but `isar` can't be null for that, and `isar` seems to always be null after 
```dart
   var isar = Isar.getInstance("messages");
```

So how can I check if Isar is already open in a background isolate? And only open it if it's not already open?

Thank you for reading all my question (I'm sorry it's so long, I'm just trying to explain as best as I can what is happening), and thank you very much for your help.

**Comment by simc** [comment, marked as answer]
You are using `Isar.getInstance` wrong. You need to provide the instance name, not the collection name.

You can specify the instance name when you open Isar using `Isar.open(name: ...)`. By default it is `"isar"`.

**Reply by tlp19** [reply]
My bad. It is indeed working now, thank you very much for your help!

**Comment by mihai-gabriel8** [comment]
hello @tlp19 

How did you manage to make isar work on background notification received?

It seems that in my case, when a notification is received but app is terminated, isar does not write to the database.

Can you help me with an example?

I am also using it like this:

```

          if (type == "chat_message") {
            var isar = Isar.getInstance();
            if (isar == null) {
              final dir = await getApplicationDocumentsDirectory();
              isar = await Isar.open(
                [
                  UserProfileSchema,
                  UserSavedNftModelSchema,
                  ChatRoomSchema,
                  MessageModelSchema,
                  CommunityMemberSchema,
                  ActivitySchema,
                ],
                directory: dir.path,
              );
              }
```

**Comment by mihai-gabriel8** [comment]
`Tried to send a platform message response, but FlutterJNI was detached from native C++. Could not send. Response ID: 14`

I get this error when I try to save a message to Isar DB when app is terminated
